engine/src/main/java/org/terasology/engine/utilities/procedural/PerlinNoise.java

```diff
@@ -72,13 +72,22 @@ public PerlinNoise(long seed, int gridDim) {
      */
     @Override
     public float noise(float posX, float posY, float posZ) {
-        int xInt = Math.floorMod(TeraMath.floorToInt(posX), permCount);
-        int yInt = Math.floorMod(TeraMath.floorToInt(posY), permCount);
-        int zInt = Math.floorMod(TeraMath.floorToInt(posZ), permCount);
 
-        float x = posX - TeraMath.fastFloor(posX);
-        float y = posY - TeraMath.fastFloor(posY);
-        float z = posZ - TeraMath.fastFloor(posZ);
+        // Domain rotation removes Perlin's characteristic square artifacts from the XZ planes, by pointing Y up the grid's main diagonal.
+        float xz = posX + posZ;
+        float s2 = xz * -0.211324865405187f;
```

Comment:
where do these "magic numbers" come from?

Reply:
They come from the vectors of the rotated coordinate space.

Ordinarily, X can be said to move in the unit vector direction <1, 0, 0>, Y in <0, 1, 0>, and Z in <0, 0, 1>. With this rotation, moving along the input for Y now moves in the unit direction <0.577, 0.577, 0.577> in the noise's internal coordinate space. Perpendicular to that, X and Z move in the directions <0.789, -0.577, -0.211> and <-0.211, -0.577, 0.789>. These vectors form a rotation matrix. The code is a simplification of the multiplication of this rotation matrix by the input coordinate, taking advantage of the many repetitions of 0.577 and the fact that 0.789 = 1-0.211.

In the past, when using this formula, I have left these values directly in the code, because the numerical choices behind the values are deeply entwined with the rotation code itself. If it would be preferable, though, I could pull out the 0.577 into a constant named something like `ROOT3_OVER_3` and the `-0.211` into `XZ_ROTATION_CONSTANT`.

Reply:
In addition to the vectors being unit length and perpendicular so as to form a rotation, the choice of <0.577, 0.577, 0.577> is because it points up the main diagonal of the noise grid, moving the XZ planes as far out of alignment with the square cube faces as possible. The <0.789, -0.577, -0.211> and <-0.211, -0.577, 0.789> are chosen out of all possibilities, because they're "the same" up to coordinate order. This reduces any slight probabilistic differences in noise behavior along those two world directions as much as possible.

Reply:
I'm actually wondering whether we should just put this explanation in a code comment here :sweat_smile: Constants help a bit (you can at least give it a name), and maybe this would be a good place to add the docstring...